compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -965,8 +959,15 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                     // dyn Auto -> dyn Auto'? ok.
                     (None, None) => Ok(CastKind::PtrPtrCast),
 
-                    // dyn Trait -> dyn Auto? should be ok, but we used to not allow it.
-                    // FIXME: allow this
+                    // dyn Trait -> dyn Auto? not ok (for now).
+                    //
+                    // As a validity invariant of pointers to trait objects, we
+                    // currently require the layout of the vtable in the metadata
+                    // to exactly match the pointee's expected vtable layout.
+                    //
+                    // Since ptr-to-ptr casts, unlike unsizing coercions, cannot change
+                    // the pointer metadata, dropping a the principal in a (non-coercion)
+                    // cast is currently not allowed.
```

Comment:
Why can they not change metadata? 

And why would you need to change it anyway? Any vtable is valid for auto traits.

Reply:
=This may have just been an oversight of mine when authoring the `dyn Trait -> dyn Auto` PR.

Reply:
> Why can they not change metadata?

It's not that they inherently cannot, they just do not as they are currently implemented and would need to be changed for this to work. 

> And why would you need to change it anyway? Any vtable is valid for auto traits.

Because our UB checks currently require the vtable to match exactly. That's what the sentence above is trying to explain.

Currently, this is considered UB and Miri will report it as such:

```rust
#![feature(core_intrinsics, custom_mir)]

trait Trait {}
impl Trait for () {}

#[custom_mir(dialect = "built")]
fn cast(ptr: *const (dyn Trait + Send)) -> *const dyn Send {
    use std::intrinsics::mir::*;
    mir! {
        {
            RET = CastPtrToPtr(ptr);
            Return()
        }
    }
}

fn main() {
    let ptr: *const (dyn Trait + Send) = &();
    let ptr = cast(ptr);
}
```

```
error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `std::marker::Send`, but encountered `Trait + std::marker::Send`
  --> src/main.rs:11:13
   |
11 |             RET = CastPtrToPtr(ptr);
   |             ^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: wrong trait in wide pointer vtable: expected `std::marker::Send`, but encountered `Trait + std::marker::Send`
   |
```

So our options here are
* changing the opsem of a PtrToPtr cast in MIR
* lowering this HIR cast to more than a single MIR cast
* declaring that the vtable must no longer match exactly (allowing `dyn Trait` vtables in `*const dyn Auto`) and adjusting the UB checks accordingly

Either way, this comment is trying to explain that we need to do more than just change this to return `Ok(CastKind::PtrPtrCast)` here.

Reply:
> =This may have just been an oversight of mine when authoring the `dyn Trait -> dyn Auto` PR.

@compiler-errors I'm not following, what oversight and PR are you talking about here?

Reply:
I'd say that we should relax the requirement. Either by allowing empty traits (`!has_own_existential_vtable_entries`) to have any vtable, or by making it actually compare vtables (something like `own_existential_vtable_entries_iter(actual).chain(repeat(Vacant)).zip(own_existential_vtable_entries_iter(expected)).all(|a, b| a == b)`)..

Reply:
Sounds plausible, but if we're going that route, then we should really make sure that upcast unsizing coercions and ptr-to-ptr casts behave consistently, so we don't up in this situation:
```rust
let ptr: *const (dyn Trait + Send) = /* ... */;

let ptr2: *const dyn Send = ptr as *const dyn Send; // unsizing coercion, changes vtable as if we unsized the original type to `dyn Send` directly
let _: *const (dyn Trait + Send) = transmute(ptr2); // UB!!, vtable doesn't match

let ptr2: *const W<dyn Send> = ptr as *const W<dyn Send>; // ptr-to-ptr cast, leaves vtable
let _: *const (dyn Trait + Send) = transmute(ptr2); // fine, vtable matches
```

Reply:
I would say that this is fine, while both casts guarantee to return a pointer with a valid vtable for `dyn Send`, there is no guarantee that it will or will not be valid for anything else.

But also, the upcast coercions don't change the vtable either.

Reply:
> But also, the upcast coercions don't change the vtable either.

They do in const eval / miri where this is currently considered UB:

```rust
trait Trait {}
impl Trait for () {}

fn main() {
    let ptr: *const (dyn Trait + Send) = &();

    let ptr2: *const dyn Send = ptr as *const dyn Send;
    let _: *const (dyn Trait + Send) = unsafe { std::mem::transmute(ptr2) };
}
```
```
error: Undefined Behavior: constructing invalid value: wrong trait in wide pointer vtable: expected `Trait + std::marker::Send`, but encountered `std::marker::Send`
 --> src/main.rs:8:49
  |
8 |     let _: *const (dyn Trait + Send) = unsafe { std::mem::transmute(ptr2) };
  |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: wrong trait in wide pointer vtable: expected `Trait + std::marker::Send`, but encountered `std::marker::Send`
  |
```

(you need `-Zextra-const-ub-checks` to trigger the check in CTFE)

> there is no guarantee that it will or will not be valid

For the sake of guarantees for stable code I think it makes sense to leave it unspecified whether this transmute is valid.

But for someone who writes unsafe code and tests it with miri I think it would be very confusing that the first case is UB, but the second case is not -- both of these are an `as` cast followed by a transmute, nothing here makes it obvious that the `as` cast actually does something different for each case.

---

Also we don't need to find a solution here: In this PR I'm just trying to add a comment to explain that there is a problem that needs to be solved before we can allow dropping the principal in non-coercion casts, and that allowing these casts won't be a one-line fix.

Reply:
I've updated the comment -- please take another look whether it is any more understandable now and/or still missing something.

also cc @RalfJung can you please double-check whether my assumptions about MIR semantics are correct here (in 46cc5e9e962fda2f99d4956ef539bf4b79fead26)? Thanks!

Reply:
What's a "non-coercion cast"? I assumed that `foo as type` will trigger coercions as required to perform the cast.

Reply:
Yes, a `T as U` cast will always trigger a coercion if possible. With "non-coercion" cast, I mean a cast that did not trigger a coercion, i.e. the case where it's not possible to coerce from `T` to `U`. Ferrocence calls these "[specialized casts](https://spec.ferrocene.dev/expressions.html#type-cast-expressions)", but I don't think we have an official term for them?